### PR TITLE
dts: xtensa: esp32: Add device tree support.

### PR DESCRIPTION
**Pull request by SavinayDharmappa**
add device tree support for esp32

Signed-off-by: Savinay Dharmappa <savinay.dharmappa@intel.com>

### CODE REVIEW

**Comment by lpereira** [comment]
Thanks for adding this! It would be nice if other devices were added to the device tree description as well, such as I2C, GPIO, pin mux, watchdog and LED PWM.

**Review by rgundi — APPROVED** [review]
LGTM

File: arch/xtensa/soc/esp32/linker.ld [hunk 1]
```diff
@@ -38,7 +39,7 @@ MEMORY
 {
   iram0_0_seg(RX): org = 0x40080000, len = 0x20000
   iram0_2_seg(RX): org = 0x400D0018, len = 0x330000
-  dram0_0_seg(RW): org = 0x3FFB0000, len = 0x50000
+  dram0_0_seg(RW): org = CONFIG_SRAM_BASE_ADDRESS, len = CONFIG_SRAM_SIZE*1K
```
Comment (andyross) [inline arch/xtensa/soc/esp32/linker.ld]: Does this bit have anything to do with DTS?  It seems dangerous to me.  I know we have those kconfig tunables, and we surely hope they're correct, but these segment addresses are the actual, true memory map of the device and can't be reasonably changed.  Allowing this to be set externally is just a footgun, no?

**Review by andyross — COMMENTED** [review]
One kibitz on the memory mapping change.

I'm not expert enough on DTS (beyond "device tree looks great") to comment much, but it looks worthwhile to me.

File: arch/xtensa/soc/esp32/linker.ld (hunk 1 above)
Comment (SavinayDharmappa) [reply arch/xtensa/soc/esp32/linker.ld]: @andyross  as a part of dts migration for all boards, we are deriving ram address from .dtsi file . In this case it is "dts/xtensa/esp32.dtsi". Where "sram0" node represent the memory mapped address of ram and it is specific to Soc.  This is the way it is done for all other arch based boards.

There is no One kibitz of memory mapping change. The dts script will generate CONFIG_SRAM_SIZE to 320. The size will be multiple of Kb.

**Comment by galak** [comment]
@SavinayDharmappa if I review, are you able to address comments?

**Comment by SavinayDharmappa** [comment]
@galak yes I will address the comments.

**Comment by nashif** [comment]
@galak waiting for your input :)

**Comment by nashif** [comment]
@galak will you provide any feedback here?

File: dts/bindings/serial/xtensa,esp32-uart.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+---
+title: ESP32 Uart
+id: xtensa,esp32-uart
```
Comment (galak) [inline dts/bindings/serial/xtensa,esp32-uart.yaml]: remove id:

File: dts/xtensa/esp32.dtsi [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+		cpu@1 {
+			device_type = "cpu";
+			compatible = "LX6";
+			reg = <1>;
+		};
+
+	};
+
+	sram0: memory@0xbe000000 {
```
Comment (galak) [inline dts/xtensa/esp32.dtsi]: remove 0x, also node address and reg don't match

File: dts/xtensa/esp32.dtsi [hunk 4]
```diff
@@ -0,0 +1,43 @@
+#include "skeleton.dtsi"
+
+/ {
+	cpus {
+		#address-cells = <1>;
+		#size-cells = <0>;
+
+		cpu@0 {
```
Comment (galak) [inline dts/xtensa/esp32.dtsi]: Do we support interrupts on xtensa?  If so we need interrupt controller node

**Review by galak — CHANGES_REQUESTED** [review]
Some minor changes and question about interrupt support.

Also do we need to update the uart driver and dts_fixup.h file?

**Comment by galak** [comment]
Any updates?

File: dts/xtensa/esp32.dtsi (hunk 4 above)
Comment (SavinayDharmappa) [reply dts/xtensa/esp32.dtsi]: No esp32 doesn't support.

File: dts/xtensa/esp32.dtsi (hunk 3 above)
Comment (SavinayDharmappa) [reply dts/xtensa/esp32.dtsi]: done

File: dts/bindings/serial/xtensa,esp32-uart.yaml (hunk 2 above)
Comment (SavinayDharmappa) [reply dts/bindings/serial/xtensa,esp32-uart.yaml]: done .

**Comment by SavinayDharmappa** [comment]
@galak i have addressed ur review comments can u review  it ?

File: dts/xtensa/esp32.dtsi [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+		cpu@1 {
+			device_type = "cpu";
+			compatible = "LX6";
+			reg = <1>;
+		};
+
+	};
+
+	sram0: memory@be000000 {
```
Comment (galak) [inline dts/xtensa/esp32.dtsi]: The Reg value and node name don't seem to match (`memory@be000000` vs `0x3FFB0000` in reg)

File: boards/xtensa/esp32/esp32.dts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+/dts-v1/;
+
+#include "esp32.dtsi"
+
+/ {
+	model = "esp32";
+	compatible = "xtensa,esp32";
+
+	aliases {
+		uart_0 = &uart0;
```
Comment (galak) [inline boards/xtensa/esp32/esp32.dts]: should be `uart-0` instead of `uart_0`

**Review by galak — CHANGES_REQUESTED** [review]
Looks good, some minor issues.

File: boards/xtensa/esp32/esp32.dts (hunk 6 above)
Comment (SavinayDharmappa) [reply boards/xtensa/esp32/esp32.dts]: done

**Review by galak — APPROVED** [review]

**Pull request merged**